### PR TITLE
Refactor syncNetworkEndpoints

### PR DESCRIPTION
**Pull request by sawsa307**
Add a waitgroup that waits for the completeness of all NEG operations.

### CODE REVIEW

File: pkg/neg/syncers/transaction.go [hunk 1]
```diff
@@ -447,43 +450,66 @@ func (s *transactionSyncer) syncNetworkEndpoints(addEndpoints, removeEndpoints m
...
 }
 
+// collectSyncResult spins off go routines to execute NEG operations
+func (s *transactionSyncer) collectSyncResult(attachBatchMap map[string][]EndpointBatch, detachBatchMap map[string][]EndpointBatch) {
+	s.logger.V(3).Info("Start collecting result.")
+
+	for zone, batches := range attachBatchMap {
+		for _, batch := range batches {
+			s.logger.V(3).Info("wg increse by one")
+			s.wg.Add(1)
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction.go:479]: I think we can greatly simplify this refactor by adding to the wait groups into attachNetworkEndpoints and detachNetworkEndpoints instead which will be marked as done in operationInternal.

Comment (swetharepakula) [reply pkg/neg/syncers/transaction.go:479]: Just realized that attach and detach NetworkEndpoints will start off more go routines so we cannot just add inside attach and detach.

Comment (sawsa307) [reply pkg/neg/syncers/transaction.go]: Thank you! As discussed offline, now collectSyncResult will only wait for all goroutine to finish and emit metrics .

File: pkg/neg/syncers/transaction.go [hunk 2]
```diff
@@ -465,25 +467,35 @@ func (s *transactionSyncer) syncNetworkEndpoints(addEndpoints, removeEndpoints m
...
 		return err
 	}
+	go s.collectSyncResult()
 	return nil
 }
 
+// collectSyncResult collects the result of the sync and emits the metrics for sync result
+func (s *transactionSyncer) collectSyncResult() {
+	s.wg.Wait()
+	s.logger.V(3).Info("Done collecting result.")
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction.go:477]: Shouldn't this publish something about the result or metric?

File: pkg/neg/syncers/transaction.go [hunk 3]
```diff
@@ -465,25 +467,34 @@ func (s *transactionSyncer) syncNetworkEndpoints(addEndpoints, removeEndpoints m
...
 	}
+	go s.collectSyncResult()
 	return nil
 }
 
+// collectSyncResult collects the result of the sync and emits the metrics for sync result
+func (s *transactionSyncer) collectSyncResult() {
+	s.wg.Wait()
+}
+
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction.go:478]: For this function, what is the proposed way to collect the results? How will you get the aggregate result of all the goroutines?

**Review by swetharepakula — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**